libcxx/include/__memory/allocator.h

```diff
@@ -76,8 +79,16 @@ struct __non_trivial_if<true, _Unique> {
 
 template <class _Tp>
 class _LIBCPP_TEMPLATE_VIS allocator : private __non_trivial_if<!is_void<_Tp>::value, allocator<_Tp> > {
-  static_assert(!is_const<_Tp>::value, "std::allocator does not support const types");
-  static_assert(!is_volatile<_Tp>::value, "std::allocator does not support volatile types");
+  static_assert(!is_const<_Tp>::value, "'std::allocator' can only allocate non-const object types");
+  static_assert(!is_volatile<_Tp>::value, "'std::allocator' can only allocate non-volatile object types");
+  static_assert(!is_reference<_Tp>::value || !is_function<typename remove_reference<_Tp>::type>::value,
+                "'std::allocator' can only allocate object types; function references are not objects (consider using "
+                "a function pointer)");
+  static_assert(!is_reference<_Tp>::value,
+                "'std::allocator' can only allocate object types; references are not objects");
```

Comment:
Could we just check `is_same<remove_cvref_t<_Tp>, _Tp>` instead?

Reply:
That erases the tailored diagnostics; noting `references are not objects`, etc., are hints I that believe significantly improves UX.

Reply:
Folks internally suggested that I change the text to the following, as they feel it'll be more helpful than what I'm currently proposing.

* `std::allocator<T const>`: `'std::allocator' cannot allocate const objects` (return to status quo)
* `std::allocator<T volatile>`: `'std::allocator' cannot allocate volatile objects` (return to status quo)
* `std::allocator<T&>`: `'std::allocator' cannot allocate references`
* `std::allocator<T()>`: `'std::allocator' cannot allocate functions`

We do lose the function reference suggestion, but I'm okay with doing that for something that's more helpful.

Reply:
Why not simply "cannot allocate cv or reference qualified objects"? I'm not sure how much of an improvement it is to specify the qualifier explicitly, since it's already shown in the diagnostic.

Reply:
I kicked the internal discussion off by asking if the average C++ programmer would be able to understand ` 'std::vector' can only allocate non-cv-qualified object types`, and the feedback I've received can be aggregated as roughly

* most C++ programmers don't know what the term "cv-qualified" means
* we should only mention relevant information, and mentioning "volatile" when `volatile` isn't used was repeatedly called out in particular
* most C++ programmers probably aren't going to know that "object" has a very different meaning to other languages
* most non-toolchain folks in the discussion seemed to prefer separate messages; toolchain folks were sort of ambivalent

What do you feel a single check will offer over independent checks?

Reply:
When you have a `const int&` you get the information that you have to remove both the `const` and the `&` at once, not only after you've removed one or the other. We also avoid three instantiations for every type we have a `vector` of. Passing this test is the common case after all.

Is cv-qualified really that uncommon/niche? When you look at the cppreference page for `const` you get basically redirected to https://en.cppreference.com/w/cpp/language/cv. FWIW you could also say "const or volatile qualified". It's a bit more verbose, but should be perfectly understandable.

Reply:
I asked in a general C++ chatroom, and ended up with feedback from managers, folks who teach C++ to Googlers, and a handful of ICs. The consensus of the discussion was that "cv-qualified" likely isn't well known, and that specific error messages will be significantly better for the user than a generic one (and especially so if "volatile" is being mentioned in the generic case: that was raised at least twice).

You do raise a good point about `int const&`, and this was also brought up in the discussion. It's certainly frustrating to [get an error, fix it, recompile, and get another error](https://youtu.be/2WZLJpMOxS4?si=9p5lIkYVQ9MhvZ1U). I'd like to address this, and also keep the specific messages. Would you be opposed to additional checks that handle these cases?

Reply:
I would be against it if that means even more classes being instantiated, but maybe I found something we'd both be happy with. Louis already wants a centralized place for the checks (which I agree would be a good idea), so maybe we could have
```c++
template <class _Tp, bool = is_same<_Tp, __remove_cvref_t<_Tp> >::value>
struct __check_container_constraints {};

template <class _Tp>
struct __check_container_constraints<_Tp, false> {
  static_assert(!is_const<_Tp>::value, "cannot allocate `const` objects");
  static_assert(!is_volatile<_Tp>::value, "Some fancy diagnostic");
  // ...
};
```
That would avoid all the instantiations in the common case. I don't care about the instantiations here if it fails anyways.

Reply:
What is your concern with respect to the number of instantiations?

Reply:
Well, I'm concerned that we instantiate a bunch of types for the rare case they are actually required for diagnostic purposes. Every instantiation takes time and memory, and for a type that is used everywhere, like `vector`, we should keep unnecessary instantiations as low as possible.

Reply:
It sounds to me like you're concerned about build times. Have you run any preliminary analysis to determine that there is a noticeable impact?

Reply:
This is more a death by a thousand cuts situation than something significant I think. If we add this many checks everywhere people will definitely notice, but not if we just add it in a single place.

Reply:
Very good point, I agree. I *think* I've got something that bridges the gap between what you suggested above, and what I was originally hoping to land.